docs/tutorial.md

```diff
@@ -1,5 +1,910 @@
+
+```js
+const StartEditingFilter = state => ({...state, editingFilter: true})
+```
+
+Actions are just functions describing transformations of the state.
+This one keep everything the same except that it sets `editingFilter` to `true`.
+
+When Hyperapp dispatches an action, it replaces the old state with the new
+one calculated using the action. Then the DOM is modified to match what the
+view returns for this new state.
+
+When `editingFilter` is true, we want to have a text input instead of a 
+span with the filter word. We can express this in the `Filter` view using a
+ternary operator (`a ? b : c`).
+
+```js
+const Filter = props => h("div", {class: "filter"}, [
+  "Filter:",
+  
+  props.editingFilter                               // <---
+  ? h("input", {type: "text", value: props.filter}) // <---
+  : h("span", {class: "filter-word"}, props.filter),
+  
+  h("button", { onClick: StartEditingFilter }, "\u270E")
+])
+```
+
+Now when you click the pencil button the text input appears.  But we still need to add
+a way to go back. We need an action to `StopEditingFilter`, and a button to dispatch it.
+
+Add the action:
+
+```js
+const StopEditingFilter = state => ({...state, editingFilter: false})
+```
+
+and update the `Filter` view again:
+
+```js
+const Filter = props => h("div", {class: "filter"}, [
+  "Filter:",
+    
+  props.editingFilter
+  ? h("input", {type: "text", value: props.filter})
+  : h("span", {class: "filter-word"}, props.filter),
+
+  props.editingFilter                                      // <---
+  ? h("button", {onClick: StopEditingFilter}, "\u2713") 
+  : h("button", {onClick: StartEditingFilter}, "\u270E"),  // <---
+])
+```
+
+When you click the pencil button, it is replaced with a check-mark button that can take you back to the first state.
+
+![Filter in edit mode](./tut2.png)
+
+
+### Capturing event-data in actions
+
+The next step is to use the input for editing the filter word. Whatever we
+type in the box should be emphasized in the story-list.
+
+Update the `Filter` view yet again:
+
+```js
+const Filter = props => h("div", {class: "filter"}, [
+  "Filter:",
+    
+  props.editingFilter
+  ? h("input", {
+    type: "text",
+    value: props.filter,
+    onInput: SetFilter,   // <----
+  })
+  : h("span", {class: "filter-word"}, props.filter),
+
+  props.editingFilter
+  ? h("button", {onClick: StopEditingFilter}, "\u2713") 
+  : h("button", {onClick: StartEditingFilter}, "\u270E"), 
+])
+```
+
+This will dispatch the `SetFilter` action everytime someone types in the input. Implement the action like this:
+
+```js
+const SetFilter = (state, event) => ({...state, filter: event.target.value})
+```
+
+The second argument to an action is known as the _payload_. Actions
+dispatched in response to an events on DOM elements receive the [event object](https://developer.mozilla.org/en-US/docs/Web/API/Event) for a payload. `event.target` refers to the input element in the DOM, and
```

Comment:
As I recently re-discovered, `event.currentTarget` might be the right thing more often than not.

Reply:
AFAIK `target` and `currentTarget` are equivalent in this case because the handler is bound to the element on which the event we're interested in will occur. Assuming that is true, I prefer `event.target` because it's short. 

I think we'll have to assume users know the difference between target and currentTarget (we can't teach javascript in the tutorial), but perhaps we should make it more clear that the event is bound to the element on which we declare `onClick`. That is perhaps not obvious to someone coming from React where (afaik) that's not how things work.